tools/pubsys/src/aws/validate_ami/mod.rs

```diff
@@ -0,0 +1,799 @@
+//! EC2 images
+
+pub(crate) mod ami;
+pub(crate) mod results;
+
+use self::ami::ImageDef;
+use self::results::{AmiValidationResult, AmiValidationResultStatus, AmiValidationResults};
+use crate::aws::client::build_client_config;
+use crate::aws::validate_ami::ami::describe_images;
+use crate::Args;
+use aws_sdk_ec2::{Client as AmiClient, Region};
+use log::{info, trace};
+use pubsys_config::InfraConfig;
+use snafu::ResultExt;
+use std::collections::{HashMap, HashSet};
+use std::fs::File;
+use std::path::PathBuf;
+use structopt::{clap, StructOpt};
+
+/// Validates EC2 images
+#[derive(Debug, StructOpt)]
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct ValidateAmiArgs {
+    /// File holding the expected amis
+    #[structopt(long, parse(from_os_str))]
+    expected_amis_path: PathBuf,
+
+    /// Optional path where the validation results should be written
+    #[structopt(long, parse(from_os_str))]
+    write_results_path: Option<PathBuf>,
+
+    #[structopt(long, requires = "write-results-path")]
+    /// Optional filter to only write validation results with these statuses to the above path
+    /// The available statuses are: `Correct`, `Incorrect`, `Missing`
+    write_results_filter: Option<Vec<AmiValidationResultStatus>>,
+
+    #[structopt(long)]
+    /// If this argument is given, print the validation results summary as a JSON object instead
+    /// of a plaintext table
+    json: bool,
+}
+
+/// Performs EC2 image validation and returns the `AmiValidationResults` object
+pub(crate) async fn validate(
+    args: &Args,
+    validate_ami_args: &ValidateAmiArgs,
+) -> Result<AmiValidationResults> {
+    info!("Parsing Infra.toml file");
+
+    // If a lock file exists, use that, otherwise use Infra.toml
+    let infra_config = InfraConfig::from_path_or_lock(&args.infra_config_path, false)
+        .context(error::ConfigSnafu)?;
+
+    let aws = infra_config.aws.clone().unwrap_or_default();
+
+    trace!("Parsed infra config: {:#?}", infra_config);
+
+    // Parse the expected ami file
+    info!("Parsing expected ami file");
+    let expected_images = parse_expected_amis(&validate_ami_args.expected_amis_path).await?;
+
+    info!("Parsed expected ami file");
+
+    // Create a HashMap of AmiClients, one for each region where validation should happen
+    let base_region = &Region::new(aws.regions[0].clone());
+    let mut ami_clients = HashMap::with_capacity(expected_images.len());
+
+    for region in expected_images.keys() {
+        let client_config = build_client_config(region, base_region, &aws).await;
+        let ami_client = AmiClient::new(&client_config);
+        ami_clients.insert(region.clone(), ami_client);
+    }
+
+    // Create a map of image_id to a bool indicating whether or not the image is public. This map
+    // is needed to determine if launch permissions should be retrieved for the image (which is the
+    // case if it is not public)
+    let expected_image_public = expected_images
```

Comment:
We are already passing expected_images into the describe_images function. Is it necessary to duplicate the image public?